shared/middleware/files_api.rb

```diff
@@ -355,4 +357,212 @@ def copy_file(endpoint, encrypted_channel_id, filename, source_filename)
 
     get_bucket_impl(endpoint).new.restore_previous_version(encrypted_channel_id, filename, request.GET['version']).to_json
   end
+
+  #
+  # GET /v3/files/<channel-id>
+  #
+  # List filenames and sizes.
+  #
+  get %r{/v3/files/([^/]+)$} do |encrypted_channel_id|
+    dont_cache
+    content_type :json
+
+    bucket = FileBucket.new
+    result = bucket.get(encrypted_channel_id, 'manifest.json', env['HTTP_IF_MODIFIED_SINCE'])
+    not_modified if result[:status] == 'NOT_MODIFIED'
+    last_modified result[:last_modified]
+
+    if result[:status] == 'NOT_FOUND'
+      { "filesVersionId": "unused", "files": [] }.to_json
+    else
+      # {
+      #   "filesVersionId": "sadfhkjahfsdj",
+      #   "files": [
+      #     {
+      #       "filename": "name.jpg",
+      #       "category": "image",
+      #       "size": 100,
+      #       "versionId": "asldfklsakdfj"
+      #     }
+      #   ]
+      # }
+      { "filesVersionId": result[:version_id], "files": JSON.load(result[:body]) }.to_json
+    end
+  end
+
+  def files_put_file(encrypted_channel_id, filename, body)
+    # read the manifest
+    bucket = FileBucket.new
+    manifest_result = bucket.get(encrypted_channel_id, 'manifest.json')
+    if manifest_result[:status] == 'NOT_FOUND'
```

Comment:
I'd love to see all the interactions with the manifest extracted into helpers (or even a helper class) so that correct use of the manifest is naturally enforced by the code.  Suppose these six lines were extracted into a static method `Manifest::load`.  Then your code might start looking like this:

```ruby
def files_put_file(encrypted_channel_id, filename, body)
  bucket = FileBucket.new
  manifest = ChannelManifest.load(bucket, encrypted_channel_id)

  # store the new file
  # ...

  manifest.upsert!(filename, new_entry_hash)
  manifest.delete!(params['delete']) if params['delete']

  new_version_id = manifest.save!(params['files-version'])

# etc...
end
```